File: lab3/task4/go_solution/t4.go
package main

import (
	"fmt"
)

// Задание 4: Игра с числами
func main() {
	fmt.Println("--- Задание 4: Игра с числами ---")
	// Структура для ключа карты, описывающая состояние игры
	type State struct {
		Rem  int // Сколько чисел осталось на столе
		Last int // Сколько чисел текущий игрок взял на своем прошлом ходу
	}

	var n, m int
	fmt.Print("Введите количество чисел (N) и ограничение на ход (M) через пробел: ")
	_, err := fmt.Scan(&n, &m)
	if err != nil {
		fmt.Println("Ошибка ввода параметров!")
		return
	}

	// Считываем сам массив чисел
	numbers := make([]int, n)
	fmt.Printf("Введите %d чисел(ла) через пробел или Enter: ", n)
	for i := 0; i < n; i++ {
		fmt.Scan(&numbers[i])
	}

	// Карта для сохранения вычисленных состояний (аналог таблицы memo)
	memo := make(map[State]bool)

	// Итеративный просчет дерева ходов снизу вверх (от 0 до N оставшихся чисел)
	for rem := 0; rem <= n; rem++ {
		for last := 0; last <= m; last++ {

			// Базовый случай: если чисел нет, текущий игрок проиграл
			if rem == 0 {
				memo[State{Rem: rem, Last: last}] = false
				continue
			}

			currentStateWin := false

			// Перебираем все варианты текущего хода (сколько чисел стереть)
			for take := 1; take <= m; take++ {
				// Проверяем правила: берем не больше, чем есть, и не повторяем свой прошлый ход
				if take <= rem && take != last {

					// Состояние, в которое попадет соперник
					nextState := State{Rem: rem - take, Last: take}

					// Если соперник из того состояния гарантированно проигрывает,
					// значит наш текущий ход — выигрышный
					if nextWin, exists := memo[nextState]; exists && !nextWin {
						currentStateWin = true
						break // Нашли победный ход, остальные варианты можно не перебирать
					}
				}
			}

			// Сохраняем результат для этой комбинации параметров
			memo[State{Rem: rem, Last: last}] = currentStateWin
		}
	}

	fmt.Println("---------------------------------------")
	fmt.Print("Результат игры (1 - Павел, 0 - Вика): ")

	// Проверяем финальное состояние игры для N элементов.
	// У Павла на первом ходу ограничений нет, поэтому Last = 0.
	initialState := State{Rem: n, Last: 0}
	if memo[initialState] {
		fmt.Println("1 (Выиграл Павел)")
	} else {
		fmt.Println("0 (Выиграла Вика)")
	}
}

File: lab3/task4/cpp_solution/t4.cpp
#include <iostream>
#include <vector>
#include <map>

using namespace std;

int main() {
    setlocale(LC_ALL, "Russian");
    
    cout << "--- Задание 4: Таблица простых чисел ---" << endl;
    int n, m;
    cout << "Введите количество чисел (N) и ограничение на ход (M) через пробел: ";
    if (!(cin >> n >> m)) {
        cout << "Ошибка ввода параметров!" << endl;
        return 1;
    }

    vector<int> numbers(n);
    cout << "Введите " << n << " чисел(ла) через пробел или Enter: ";
    for (int i = 0; i < n; i++) {
        cin >> numbers[i];
    }

    map<pair<int, int>, bool> memo;

    for (int rem = 0; rem <= n; ++rem) {
        for (int last = 0; last <= m; ++last) {
            if (rem == 0) {
                memo[{rem, last}] = false;
                continue;
            }

            bool current_state_win = false;

            for (int take = 1; take <= m; ++take) {
                if (take <= rem && take != last) {
                    pair<int, int> next_state = {rem - take, take};
                    if (memo.count(next_state) && !memo[next_state]) {
                        current_state_win = true;
                        break;
                    }
                }
            }
            memo[{rem, last}] = current_state_win;
        }
    }

    cout << "---------------------------------------" << endl;
    cout << "Результат игры (1 - Павел, 0 - Вика): ";

    if (memo[{n, 0}]) {
        cout << 1 << " (Выиграл Павел)" << endl;
    } else {
        cout << 0 << " (Выиграла Вика)" << endl;
    }
    
    return 0;
}